apollo-backend/src/main/java/io/logz/apollo/services/SlaveService.java

```diff
@@ -0,0 +1,143 @@
+            throw new RuntimeException("Could not understand slaves params");
+        }
+
+        slaveId = UUID.randomUUID().toString();
+        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
+    }
+
+    @PostConstruct
+    public void start() {
+        cleanupUnusedSlaves();
+
+        if (isSlave) {
+            claimSlave();
+            scheduledExecutorService.scheduleWithFixedDelay(this::keepAlive,
+                    0, apolloConfiguration.getSlave().getKeepaliveIntervalSeconds(),
+                    TimeUnit.SECONDS);
+
+            started = true;
+        }
+    }
+
+    @PreDestroy
+    public void stop() {
+        started = false;
+        scheduledExecutorService.shutdownNow();
+        slaveDao.getAllSlaves()
+                .stream()
+                .filter(slave -> slave.getSlaveId().equals(slaveId))
+                .findFirst()
+                .ifPresent(slave -> slaveDao.removeAllSlavesById(slave.getSlaveId()));
+    }
+
+    public Boolean getSlave() {
+        return isSlave;
+    }
+
+    public List<Integer> getEnvironmentIds() {
+        return environmentIds;
+    }
+
+    public List<Integer> getAllValidSlavesEnvironmentIds() {
+        return slaveDao.getAllSlaves().stream()
+                .filter(slave -> slave.getSecondsSinceLastKeepalive()
+                        >= apolloConfiguration.getSlave().getKeepaliveIntervalSeconds() * 2)
+                .map(Slave::getEnvironmentId)
+                .collect(Collectors.toList());
+    }
+
+    public boolean isStarted() {
+        return started;
+    }
+
+    private void claimSlave() {
+        environmentIds.forEach(environmentId -> {
+            Slave slave = new Slave();
+            slave.setSlaveId(slaveId);
+            slave.setEnvironmentId(environmentId);
+            slave.setLastKeepalive(new Date());
+            slaveDao.addSlave(slave);
+        });
+    }
+
+    private void keepAlive() {
+        slaveDao.getAllSlaves()
+                .stream()
+                .filter(slave -> slave.getSlaveId().equals(slaveId))
+                .forEach(slave -> slaveDao.keepalive(slave.getSlaveId()));
+    }
+
+    private List<Integer> parseEnvironmentIds() {
+        try {
+            String envVar = getEnvVarOrProperty(SLAVE_CSV_ENVIRONMENTS);
+            if (envVar == null) {
+                return null;
+            }
+
+            return Splitter.on(",").trimResults()
+                    .splitToList(envVar)
+                    .stream().map(Integer::parseInt)
+                    .collect(Collectors.toList());
+
+        } catch (NumberFormatException e) {
+            logger.error("Could not parse int list from {}", SLAVE_CSV_ENVIRONMENTS);
+            return null;
+        }
+    }
+
+    private void cleanupUnusedSlaves() {
+        slaveDao.getAllSlaves().stream().filter(slave -> slave.getSecondsSinceLastKeepalive() >=
```

Comment:
Create a new DAO method `cleanupDeadSlaves(@Param("oldestValidKeepAlive") int oldestValidKeepalive)` with the value of `apolloConfiguration.getSlave().getKeepaliveIntervalSeconds() * 4`

Reply:
There's an issue here, let's say we have 3 environments, 1 which for some reason we didn't update the keepalive and 2 that we did, here you'll remove the records of all 3 of them